### ISSUE TITLE
`tfe_workspace` `vcs_repo` error

### ISSUE THREAD

**Issue by bensojona** [terraform-providers/terraform-provider-tfe#9]
It's not quite clear how [this](https://www.terraform.io/docs/providers/tfe/r/workspace.html#vcs_repo) should work. My assumption is that it assumes you've already got a VCS connection on the org setup, but even when that's the case, I'm not sure how to configure it so it knows which VCS connection to use.

I get the below error when running a plan. Not sure if that's a bug, either way, a more descriptive error could help.

```
Error: Error applying plan:

1 error(s) occurred:

* tfe_workspace.producer: 1 error(s) occurred:

* tfe_workspace.producer: Error updating workspace dynamic-aws-creds-producer for organization jbenson_test: internal server error
```

Config looks like this..

```
resource "tfe_workspace" "producer" {
  name         = "${var.producer_name}"
  organization = "${tfe_organization.org.id}"
  working_directory = "${var.producer_wd}"

  vcs_repo {
    identifier = "hashicorp/terraform-guides"
    branch     = "f-dynamic-aws-creds-tfe"
    oauth_token_id = "${var.token}"
  }
}
```

A nice feature enhancement would be to be able to create a VCS connection on the org as a resource as well.

**Comment by djaboxx** [comment]
I've actually created a new resource for this provider that does create the new VCS (only for github currently). You can see it here. 

https://github.com/HappyPathway/terraform-provider-tfe/blob/master/tfe/resource_tfe_organization_vcs.go. 

This requires a new go-tfe client library as well, which is here: 
https://github.com/HappyPathway/go-tfe/blob/master/oauth_client.go

But, even with this, you still have to add the Callback URL to your Oauth App in github. The transaction seems pretty difficult to automate as it doesn't seem that there's any way to modify the Oauth Apps through the Github API.

**Comment by svanharmelen** [comment]
Hi @bensojona! When I try your exact config, it seems to work as expected for me, so let's see if we can troubleshoot this a bit to find the problem...

1. Can you confirm that the VCS connection already exists and that you can create a new workspace with the same settings manually?
2. I noticed the error said `Error updating workspace`, so was this workspace initially created with Terraform as well?

Hope this will help us understand what is causing the error.

> A nice feature enhancement would be to be able to create a VCS connection on the org as a resource as well.

That is definitely on the roadmap for the next release!

**Comment by bensojona** [comment]
> Can you confirm that the VCS connection already exists and that you can create a new workspace with the same settings manually?

The VCS connection did not exist, that was my fault, forgot I had created a new org vs. using an existing. Though after creating the VCS connection, I still got an error. 

> I noticed the error said Error updating workspace, so was this workspace initially created with Terraform as well?

It was! I did the below steps.

1. Created the workspace with TFE provider upon initial apply
1. Tried to add the `vcs_repo` block to that workspace after - it errored (first error below) because I didn't have the VCS connection setup
1. Added the VCS connection to the org
1. Tained the workspace so it would re-create with the proper VCS connection
1. Tried to apply again and it errored (second error below)

## Step 2 Apply Error
```
Error: Error applying plan:

1 error(s) occurred:

* tfe_workspace.producer: 1 error(s) occurred:

* tfe_workspace.producer: Error updating workspace dynamic-aws-creds-producer for organization jbenson_test: internal server error
```

## Step 5 Apply Error
```
Error: Error applying plan:

1 error(s) occurred:

* tfe_workspace.producer: 1 error(s) occurred:

* tfe_workspace.producer: Error creating workspace dynamic-aws-creds-producer for organization jbenson_test: Invalid Attribute OAuth token does not exist
```

**Comment by svanharmelen** [comment]
Ah, but that is a clear error: `Invalid Attribute OAuth token does not exist` So what token ID did you use and/or how did you get it? As currently that is a bit nasty to get...

**Comment by djaboxx** [comment]
that script that I posted in the gist helps with getting the correct token. Though, I believe it will only work if you have one VCS connection in the Org.

**Comment by bensojona** [comment]
@djaboxx @svanharmelen oh shoot, I'm sorry, did not read thoroughly! I just created an OAuth token via Github. Let me give your method a shot @djaboxx

**Comment by bensojona** [comment]
@svanharmelen - the script @djaboxx provided did the trick!

**Comment by svanharmelen** [comment]
Cool... Thanks for your help @djaboxx!

And as mentioned we are working on improving this in a next release, to make it easier to get the correct token.

**Comment by svanharmelen** [comment]
@pgrinstead1 at the moment this is hard as a token does not have a way to be uniquely identified. But if you only have one VCS connection, it's a bit easier to find the correct ID.

For now I suggest you open the following URL in your browser (when logged into PTFE, so with an authenticated session): https://ptfe.company.com/api/v2/organizations/your-org-name/oauth-clients

This will return a JSON document containing a list of all your defined VCS connections with their attributes and `relationships`. In the `relationships` section there will be an `oauth-tokens` entry which contains the token ID.

We understand that this is far from perfect, but for now this is probably the best we can do.

Another solution would be to create a new VCS connection using the `tfe_oauth_client` resource. It's fine to have multiple VCS connections to the same VCS service so you could add the new connection next to the existing connection and then reference the exported `oauth_token_id` in resources that need the ID (e.g. `tfe_workspace`).

I hope this helps solve the issue for now?

**Comment by pgrinstead1** [comment]
@svanharmelen your suggestion worked like a charm. What a wonky thing. Thank you so much!

**Comment by pawelprazak** [comment]
Thanks to @svanharmelen suggestion to just use the browser and @djaboxx explanation of the mechanism I was able to hack around the limitations, but it would be awesome if this would be easier, or at least documented in the provider docs.

Tha main confusion is regarding how to get the `oauth_token_id` and what it is exactly.

For anyone that needs to use a shell, try this:
```
curl -v -H "Authorization: Bearer <redacted>" https://ap https://app.terraform.io/api/v2/organizations/<redacted>/oauth-tokens | jq .data[].id
```

**Comment by pawelprazak** [comment]
For anyone trying to get it working, here is what worked for me:

[Create an oauth client](https://www.terraform.io/docs/cloud/api/oauth-clients.html#create-an-oauth-client):
```
resource "tfe_oauth_client" "github" {
  organization     = tfe_organization.org.name
  api_url          = "https://api.github.com"
  http_url         = "https://github.com"
  service_provider = "github"
  oauth_token      = var.github_oauth_token
}
```

Where `var.github_oauth_token` i a [personal access token](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token) to [a machine user](https://docs.github.com/en/free-pro-team@latest/developers/overview/managing-deploy-keys#machine-users).

I gave the machine user's personal access token the following scopes: `admin:repo_hook, repo` and an owner permissions over the organization (sadly couldn't figure out how to narrow down access for the user itself).

Then used it in the workspace definition:
```
resource "tfe_workspace" "test" {
  name         = "test"
  organization = tfe_organization.org.name
  vcs_repo {
    identifier     = "<redacted>/test-tf"
    oauth_token_id = tfe_oauth_client.github.oauth_token_id
  }
}
```

**Comment by theorange7** [comment]
I tried using `data "tfe_oauth_client" "client_id"` on our self hosted TFE which is connected to Gitlab. On the docs [here](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs/data-sources/oauth_client) it shows an example with the token id with "oc" prefix. However, when I look at the settings on our TFE instance, all the VCS provider tokens begin with "ot". 

Hence it is within expectations that data.tfe_oauth_client.client_id fails.. I'd like to understand why. What exactly does "oc" refer to? From the above comment, does it mean this data block only supports Personal Access tokens?

On a positive note, I did manage to configure a tfe workspace using the `tfe_workspace` resource by passing the oauth token id directly instead of using the data block to first retrieve it.

TLDR: What is the difference between "oc" and "ot" tokens? What do the prefixes mean? How should data.tfe_oauth_client be used exactly?

**Comment by m0ar** [comment]
This was still foggy for me, so just making it super clear how to declare the `tfe_oauth_client` data source:

```terraform
data "tfe_oauth_client" "client" {
  # In TF Cloud under VCS Providers, you find the OAuth Token ID starting with "ot-"
  # Now go to https://app.terraform.io/api/v2/organizations/<Your org>/oauth-tokens
  # There you will find a relationship to an OAuth client with an ID starting with "oc-"
  oauth_client_id = "oc-<redacted>"
}
```
